doc/kernel/drivers/index.rst

```diff
@@ -122,36 +125,42 @@ Most drivers will be implementing a device-independent subsystem API.
 Applications can simply program to that generic API, and application
 code is not specific to any particular driver implementation.
 
+If all driver API instances are assigned to their respective API linker section
+use :c:macro:`DEVICE_API_IS()` to verify the API's type.
+
 A subsystem API definition typically looks like this:
 
 .. code-block:: C
 
   typedef int (*subsystem_do_this_t)(const struct device *dev, int foo, int bar);
   typedef void (*subsystem_do_that_t)(const struct device *dev, void *baz);
 
-  struct subsystem_api {
+  __subsystem struct subsystem_driver_api {
         subsystem_do_this_t do_this;
         subsystem_do_that_t do_that;
   };
 
   static inline int subsystem_do_this(const struct device *dev, int foo, int bar)
   {
-        struct subsystem_api *api;
+        CHECKIF(!DEVICE_API_IS(subsystem, dev)) {
+              return -ENODEV;
+        }
 
-        api = (struct subsystem_api *)dev->api;
-        return api->do_this(dev, foo, bar);
+        return DEVICE_API_GET(subsystem, dev)->do_this(dev, foo, bar);
```

Comment:
This is just hiding that everything is a void*, which again wouldn't actually be a void* if we had proper typing of the devices

Reply:
This isn't necessarily type checking per-se (any pointer can be cast to one of the expected type to get around type checking at the compiler level).

It's more like assuring that a Thing is in fact a Thing by checking that a given Thing pointer is a member of the array of all Things, which is pretty hard to argue with (in terms of static RTOS behaviour).

Drivers loaded at runtime really should also get the same kind of check as a kind of minimal security practice.

Currently with userspace, the kernel validates all device API pointer arguments as being readable or writeable memory regions. A similar check is done for permissions on kernel objects. Somehow we completely bypass any checks whatsoever for (arguably) the most important pointer in a device API call (the `const struct device *dev`), which makes it trivial to trigger faults (or worse - not trigger faults but have unpredictable side-effects) in syscall context. 

Do we already support drivers loaded at runtime? If so, I think that means they implicitly rely on this (IMHO) hole in security.

Reply:
> It's more like ensuring that a Thing is in fact a Thing by checking that a given Thing pointer is a member of the array of all Things, which is pretty hard to argue with (in terms of static RTOS behaviour).

Right, which could be done by the compiler if we had proper types instead of void*'s, which is the right place to do this. Unless your talking about userspace where you have untrusted code perhaps? In that case sure, but then this sort of thing should be done in z_vrfy?

Reply:
Definitely userspace, and agree about z_vrfy.

Reply:
Regarding runtime loadable drivers, we are already boxed in quite substantionally. We don't have a way of storing the devicetree in ROM, so drivers have to be compiled in, and "switched between" at runtime. Given that these drivers are compiled in, their APIs would be placed in the correct API sections as well :) Unless by runtime loadable you mean "truly" runtime loadable, I don't see this PR as any more of a box :)

Reply:
To be clear here, my main complaint is this is kicking the can down the road of using proper types a compiler will yell about and instead leaving this to a runtime check that I guess might blow up at any random time while firmware is running.

Reply:
I see this PR as an improvement to the status quo, not really a new feature. Kicking the can down the road would IMO imply we are breaking the status quo with an incomplete, short-term solution, which I just don't think describes this PR.